# Water Crisis Analysis in Maji Ndogo – Part 2: Data Cleaning & Transformation

## Project Overview
This notebook is the second part of a four-part SQL portfolio project focused on analysing water access, infrastructure, and quality in Maji Ndogo. The objective of this phase is to prepare the dataset for reliable analysis by identifying, cleaning, and transforming inconsistencies across multiple tables.
Building on the exploratory insights from Part 1, this stage focuses on improving data quality and ensuring that the dataset is structurally and logically sound for deeper analysis and decision-making.

## Objectives
- Clean and standardise key fields across tables (e.g., employee records, phone numbers)
- Identify and resolve data inconsistencies and formatting issues
- Transform raw data into structured, analysis-ready formats
- Validate assumptions made during exploratory analysis
- Prepare datasets for downstream analysis and reporting

## Key Areas Covered
- String manipulation and formatting (REPLACE, LOWER, CONCAT, TRIM)
- Data validation and correction using SQL UPDATE statements
- Aggregations and grouping for structured insights
- Handling inconsistent or misleading data representations
- Preparing derived fields for analysis

## Project Structure
This project is structured into four parts:

1. **Exploratory Data Analysis (Part 1)** – Understanding the dataset  
2. **Data Cleaning & Transformation (Part 2)** – Preparing data *(this notebook)*  
3. **Data Validation & Auditing (Part 3)** – Ensuring data integrity  
4. **Analysis & Insights (Part 4)** – Generating actionable recommendations  

### Data Source Note
This project is inspired by concepts from the ALX Data Analyst Programme.  
While guided by the programme structure, all SQL queries and analysis were implemented independently.

The dataset is not included in this repository; however, a data dictionary is provided for reference.

In [1]:
import pandas as pd
from getpass import getpass

In [2]:
%load_ext sql

In [3]:
password = getpass('Enter DB password: ')

%sql mysql+pymysql://root:{password}@localhost:3309/md_water_services

Enter DB password:  ········


'Connected: root@md_water_services'

In [4]:
%%sql

SHOW TABLES;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
8 rows affected.


Tables_in_md_water_services
data_dictionary
employee
global_water_access
location
visits
water_quality
water_source
well_pollution


In [5]:
%%sql

SELECT *
FROM employee
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
10 rows affected.


assigned_employee_id,employee_name,phone_number,email,address,province_name,town_name,position
0,Amara Jengo,+99637993287,None,36 Pwani Mchangani Road,Sokoto,Ilanga,Field Surveyor
1,Bello Azibo,+99643864786,None,129 Ziwa La Kioo Road,Kilimani,Rural,Field Surveyor
2,Bakari Iniko,+99222599041,None,18 Mlima Tazama Avenue,Hawassa,Rural,Field Surveyor
3,Malachi Mavuso,+99945849900,None,100 Mogadishu Road,Akatsi,Lusaka,Field Surveyor
4,Cheche Buhle,+99381679640,None,1 Savanna Street,Akatsi,Rural,Field Surveyor
5,Zuriel Matembo,+99034075111,None,26 Bahari Ya Faraja Road,Kilimani,Rural,Field Surveyor
6,Deka Osumare,+99379364631,None,104 Kenyatta Street,Akatsi,Rural,Field Surveyor
7,Lalitha Kaburi,+99681623240,None,145 Sungura Amanpour Road,Kilimani,Rural,Field Surveyor
8,Enitan Zuri,+99248509202,None,117 Kampala Road,Hawassa,Zanzibar,Field Surveyor
10,Farai Nia,+99570082739,None,33 Ang??lique Kidjo Avenue,Amanzi,Dahabu,Field Surveyor


## Data Cleaning: Employee Records
The employee table contains key personnel information but includes missing and inconsistent fields. To support communication workflows, email addresses are generated using a standardised format "first_name.last_name@ndogowater.gov".

This section focuses on:
- Creating structured email addresses from employee names
- Ensuring consistent formatting across records
- Updating the database safely using validated transformations

In [6]:
%%sql

SELECT 
    CONCAT(LOWER(REPLACE(employee_name, ' ', '.')), '@ndogowater.gov') AS new_email2
FROM employee
LIMIT 4;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
4 rows affected.


new_email2
amara.jengo@ndogowater.gov
bello.azibo@ndogowater.gov
bakari.iniko@ndogowater.gov
malachi.mavuso@ndogowater.gov


In [7]:
%%sql

UPDATE employee
SET email = CONCAT(LOWER(REPLACE(employee_name, ' ', '.')), '@ndogowater.gov') 

 * mysql+pymysql://root:***@localhost:3309/md_water_services
56 rows affected.


[]

## Data Standardisation: Phone Numbers

Initial inspection revealed inconsistencies in phone number formatting, specifically trailing spaces that affect data integrity. The phone numbers should be 12 characters long, consisting of the plus sign, area code (99), and the phone number digits. However, when the LENGTH(column) is applied, it returns 13 characters, indicating there's an extra character. 

To ensure compatibility with downstream systems (e.g., messaging services), string cleaning techniques are applied to:
- Remove leading and trailing spaces
- Enforce consistent length and format validation

In [8]:
%%sql
SELECT
LENGTH(phone_number)
FROM
employee
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
5 rows affected.


LENGTH(phone_number)
13
13
13
13
13


In [9]:
%%sql

SELECT Rtrim(phone_number) AS number_trimmed
FROM employee
LIMIT 3;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
3 rows affected.


number_trimmed
+99637993287
+99643864786
+99222599041


In [10]:
%%sql

UPDATE employee
SET phone_number = Rtrim(phone_number)

 * mysql+pymysql://root:***@localhost:3309/md_water_services
56 rows affected.


[]

In [11]:
%%sql

SELECT
LENGTH(phone_number)
FROM
employee
LIMIT 4;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
4 rows affected.


LENGTH(phone_number)
12
12
12
12


## Exploratory Aggregation: Employee Distribution
To better understand workforce allocation, employee records are aggregated by location.
This provides insight into:
- Geographic distribution of field workers
- Workforce concentration in rural vs urban areas

In [12]:
%%sql

SELECT town_name, COUNT(*) AS employee_count
FROM employee
GROUP BY town_name
ORDER BY employee_count DESC, town_name;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
9 rows affected.


town_name,employee_count
Rural,29
Dahabu,6
Harare,5
Lusaka,4
Zanzibar,4
Ilanga,3
Serowe,3
Kintampo,1
Yaounde,1


## Performance Analysis: Field Surveyors
To recognise contributions and assess operational effort, the number of visits conducted by each surveyor is analysed.
This section identifies:
- Top-performing field surveyors based on activity volume
- Supporting contact details for recognition and reporting

In [13]:
%%sql

SELECT assigned_employee_id, COUNT(*) as visit_count
FROM visits
GROUP BY assigned_employee_id
ORDER BY visit_count DESC
LIMIT 3;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
3 rows affected.


assigned_employee_id,visit_count
1,3708
30,3676
34,3539


In [14]:
%%sql

SELECT assigned_employee_id,employee_name,phone_number,email	
FROM employee
WHERE assigned_employee_id IN (1, 30, 34);

 * mysql+pymysql://root:***@localhost:3309/md_water_services
3 rows affected.


assigned_employee_id,employee_name,phone_number,email
1,Bello Azibo,+99643864786,bello.azibo@ndogowater.gov
30,Pili Zola,+99822478933,pili.zola@ndogowater.gov
34,Rudo Imani,+99046972648,rudo.imani@ndogowater.gov


## Location Analysis: Distribution of Water Sources
Understanding the geographical distribution of water sources is critical for planning interventions.
This analysis examines:
- Distribution across towns and provinces
- Representation of rural vs urban areas
- Coverage consistency across regions

In [15]:
%%sql

SELECT town_name, COUNT(*) AS records_per_town
FROM location
GROUP BY town_name 
ORDER BY records_per_town DESC
LIMIT 3;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
3 rows affected.


town_name,records_per_town
Rural,23740
Harare,1650
Amina,1090


In [16]:
%%sql
-- count the records per province
SELECT province_name, COUNT(*) AS records_per_province
FROM location
GROUP BY province_name
ORDER BY records_per_province DESC;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
5 rows affected.


province_name,records_per_province
Kilimani,9510
Akatsi,8940
Sokoto,8220
Amanzi,6950
Hawassa,6030


From the table above, it's pretty clear that most of the water sources in the survey are situated in small rural communities, scattered across Maji Ndogo.
If we count the records for each province, most of them have a similar number of sources, so every province is well-represented in the survey.

In [17]:
%%sql

SELECT town_name, province_name, COUNT(*) AS records_per_town
FROM location
GROUP BY town_name, province_name
ORDER BY province_name, records_per_town DESC;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
31 rows affected.


town_name,province_name,records_per_town
Rural,Akatsi,6290
Lusaka,Akatsi,1070
Harare,Akatsi,800
Kintampo,Akatsi,780
Rural,Amanzi,3100
Asmara,Amanzi,930
Dahabu,Amanzi,930
Amina,Amanzi,670
Pwani,Amanzi,520
Abidjan,Amanzi,400


These results above infer us that the field surveyors did an excellent job of documenting the status of our country's water crisis. Every province and town has many documented sources. This increases confidence that the data is most likely reliable enough to base our decisions on.

In [18]:
%%sql

SELECT *
FROM location
LIMIT 2;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
2 rows affected.


location_id,address,province_name,town_name,location_type
AkHa00000,2 Addis Ababa Road,Akatsi,Harare,Urban
AkHa00001,10 Addis Ababa Road,Akatsi,Harare,Urban


In [19]:
%%sql

SELECT location_type, COUNT(*) AS num_sources
FROM location
GROUP BY location_type
LIMIT 2;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
2 rows affected.


location_type,num_sources
Urban,15910
Rural,23740


In [20]:
%%sql

SELECT 23740 / (15910 + 23740) * 100

 * mysql+pymysql://root:***@localhost:3309/md_water_services
1 rows affected.


23740 / (15910 + 23740) * 100
59.8739


***Result shows that 60% of all water sources in the data set are in rural communities.***

## Water Source Analysis
This section explores the types and usage of water sources across Maji Ndogo.
Key questions addressed:
- Distribution of water source types
- Population served per source
- Average load per infrastructure type
Special attention is given to interpreting aggregated records correctly, particularly for household-level infrastructure.

In [21]:
%%sql

SELECT COUNT(*)
FROM water_source
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
1 rows affected.


COUNT(*)
39650


In [22]:
%%sql

SELECT *
FROM water_source
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
10 rows affected.


source_id,type_of_water_source,number_of_people_served
AkHa00000224,tap_in_home,956
AkHa00001224,tap_in_home_broken,930
AkHa00002224,tap_in_home_broken,486
AkHa00003224,well,364
AkHa00004224,tap_in_home_broken,942
AkHa00005224,tap_in_home,736
AkHa00006224,tap_in_home,882
AkHa00007224,tap_in_home,554
AkHa00008224,well,398
AkHa00009224,well,346


In [23]:
%%sql

SELECT type_of_water_source, COUNT(*) as number_of_sources
FROM water_source
GROUP BY type_of_water_source;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
5 rows affected.


type_of_water_source,number_of_sources
tap_in_home,7265
tap_in_home_broken,5856
well,17383
shared_tap,5767
river,3379


In [24]:
%%sql

SELECT type_of_water_source,
       ROUND(AVG(number_of_people_served), 0) AS avg_people_served
FROM water_source
GROUP BY type_of_water_source
ORDER BY avg_people_served DESC;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
5 rows affected.


type_of_water_source,avg_people_served
shared_tap,2071
river,699
tap_in_home_broken,649
tap_in_home,644
well,279


### Analytical Consideration
It is important to interpret aggregated values carefully. Records representing household taps combine multiple households into a single entry, as noted in Part 1 of the analysis. For instance, the calculated average of 644 people per `tap_in_home` does not imply that this number of individuals share a single tap. Instead, each record aggregates several households, with an average of approximately six people per household. This suggests that one `tap_in_home` record represents roughly 100 individual household taps (644 ÷ 6 ≈ 100).
Therefore, calculated averages must be contextualised to avoid misleading conclusions and to ensure accurate decision-making. Proper interpretation of these values is essential when assessing infrastructure load and prioritising interventions across water source types.

In [25]:
%%sql

SELECT type_of_water_source , SUM(number_of_people_served) as population_served
FROM water_source
GROUP BY type_of_water_source
ORDER BY population_served DESC;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
5 rows affected.


type_of_water_source,population_served
shared_tap,11945272
well,4841724
tap_in_home,4678880
tap_in_home_broken,3799720
river,2362544


In [26]:
%%sql

SELECT type_of_water_source,
       ROUND(SUM(number_of_people_served) *100/(SELECT SUM(number_of_people_served) FROM water_source), 0) AS percentage_served
FROM water_source
GROUP BY type_of_water_source
ORDER BY percentage_served DESC;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
5 rows affected.


type_of_water_source,percentage_served
shared_tap,43
well,18
tap_in_home,17
tap_in_home_broken,14
river,9


In [27]:
%%sql
SELECT type_of_water_source,
    SUM(number_of_people_served) AS population_served,
    ROUND(
        SUM(number_of_people_served) * 100.0 / 
        (SELECT SUM(number_of_people_served) FROM water_source),
        2
    ) AS percentage_served
FROM water_source
GROUP BY type_of_water_source
ORDER BY percentage_served DESC;


 * mysql+pymysql://root:***@localhost:3309/md_water_services
5 rows affected.


type_of_water_source,population_served,percentage_served
shared_tap,11945272,43.24
well,4841724,17.52
tap_in_home,4678880,16.94
tap_in_home_broken,3799720,13.75
river,2362544,8.55


Approximately 43% of the population relies on shared taps, with an average of around 2,000 people served per shared tap, indicating significant pressure on this infrastructure. Combining `tap_in_home` and `tap_in_home_broken`, approximately 31% of the population has water infrastructure installed within their homes. However, a substantial portion of this infrastructure, pproximately 45% (14 out of 31) is non-functional. It is important to note that the issue is not with the taps themselves, but with the supporting systems such as treatment plants, reservoirs, pipelines, and pumps.
Additionally, 18% of the population depends on wells. Of these, only 4,916 out of 17,383 wells are classified as clean, representing just 28%, which highlights a significant concern regarding water quality.

## Prioritisation Framework

To support data-driven decision-making, water sources are ranked based on the population they serve. tap_in_home is ommitted from the ranking since it is anyways the best source available
This enables:
- Identification of high-impact intervention points
- Prioritisation of infrastructure improvements
- Efficient allocation of resources

In [28]:
%%sql

SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS total_people_served,
    RANK() OVER (ORDER BY SUM(number_of_people_served) DESC) AS rank_by_population
FROM water_source
WHERE type_of_water_source != 'tap_in_home'
GROUP BY type_of_water_source
ORDER BY rank_by_population;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
4 rows affected.


type_of_water_source,total_people_served,rank_by_population
shared_tap,11945272,1
well,4841724,2
tap_in_home_broken,3799720,3
river,2362544,4


Based on these findings, shared taps should be prioritised for intervention, followed by wells and other water sources. The next step is to determine which specific locations require the most urgent attention. Applying the same prioritisation logic, water sources that serve the largest populations should be addressed first to maximise the impact of improvements.

In [29]:
%%sql

SELECT source_id,
    type_of_water_source,
    number_of_people_served,
    RANK() OVER (ORDER BY number_of_people_served DESC) AS priority_rank
FROM water_source
WHERE type_of_water_source IN ('shared_tap', 'well')
ORDER BY priority_rank
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
10 rows affected.


source_id,type_of_water_source,number_of_people_served,priority_rank
HaRu19509224,shared_tap,3998,1
AkRu05603224,shared_tap,3998,1
AkRu04862224,shared_tap,3996,3
AmAs10911224,shared_tap,3996,3
KiHa22867224,shared_tap,3996,3
HaRu19839224,shared_tap,3994,6
KiZu31330224,shared_tap,3994,6
KiRu28630224,shared_tap,3992,8
KiZu31415224,shared_tap,3992,8
KiRu26218224,shared_tap,3990,10


Ok, so we should fix shared taps first, then wells, and so on. But the next question is, which shared taps or wells should be fixed first? We can use
the same logic; the most used sources should really be fixed first.

In [30]:
%%sql
-- Similar to the previous code, however, it uses DENSE_RANK instead of RANK
SELECT source_id,
    type_of_water_source,
    number_of_people_served,
    DENSE_RANK() OVER (ORDER BY number_of_people_served DESC) AS priority_rank
FROM water_source
WHERE type_of_water_source IN ('shared_tap', 'well')
ORDER BY priority_rank
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
5 rows affected.


source_id,type_of_water_source,number_of_people_served,priority_rank
HaRu19509224,shared_tap,3998,1
AkRu05603224,shared_tap,3998,1
AkRu04862224,shared_tap,3996,2
AmAs10911224,shared_tap,3996,2
KiHa22867224,shared_tap,3996,2


## Queue Time Analysis
The visits dataset provides insights into water access challenges through queue time measurements.

This analysis focuses on:
- Average queue durations
- Temporal patterns (day of week, time of day)
- Identifying peak demand periods
Advanced SQL techniques such as DateTime functions and conditional aggregation are used to uncover patterns in water collection behaviour.

In [31]:
%%sql
-- This query shows how long the survey took

SELECT MIN(time_of_record) as survey_start,
        MAX(time_of_record) as survey_end, 
        DATEDIFF(MAX(time_of_record), MIN(time_of_record) ) AS survey_duration_days
FROM visits;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
1 rows affected.


survey_start,survey_end,survey_duration_days
2021-01-01 09:10:00,2023-07-14 13:53:00,924


The survey was conducted over 924 days, approximately 2.5 years, reflecting a significant and sustained data collection effort.

To better understand access challenges, the next step is to evaluate the average time people spend in queues across Maji Ndogo.
It is important to note that certain water sources, such as `tap_in_home`, do not involve queueing and are recorded with a value of 0 in the `time_in_queue` column. To ensure an accurate calculation of average queue times, these values is excluded.

In [32]:
%%sql
SELECT AVG(NULLIF(time_in_queue, 0)) AS avg_queue_time
FROM visits;

 * mysql+pymysql://root:***@localhost:3309/md_water_services
1 rows affected.


avg_queue_time
123.2574


The analysis indicates an average queue time of approximately 123 minutes. This represents a substantial waiting period of two hoursfor individuals without access to in-home water infrastructure.

While this average may initially appear manageable, it likely masks variability across different days and locations. In practice, demand for water is often concentrated during specific periods typically in the early mornings or late evening on weekdays or even weekends when most people are not at work or school, meaning some individuals may experience significantly longer wait times. This suggests that access is not only time-consuming but also inconsistent, placing an additional burden on affected communities.

In [33]:
%%sql

SELECT ROUND(AVG(time_in_queue)) AS avg_queue_time,
       DAYNAME(time_of_record) AS day_of_week
FROM visits
WHERE NULLIF(time_in_queue, 0) IS NOT NULL
GROUP BY DAYNAME(time_of_record)
ORDER BY avg_queue_time DESC;


 * mysql+pymysql://root:***@localhost:3309/md_water_services
7 rows affected.


avg_queue_time,day_of_week
246,Saturday
137,Monday
120,Friday
108,Tuesday
105,Thursday
97,Wednesday
82,Sunday


The analysis shows that average queue times are highest on Saturdays, indicating increased demand for water access on that day. In contrast, Sundays have the lowest average queue times.

This pattern may suggest the influence of social or cultural factors on water collection behaviour. However, further contextual data would be required to draw definitive conclusions regarding the underlying causes.


### Temporal Analysis

To further understand usage patterns, it is important to examine how water collection varies throughout the day. The next step is to analyse the distribution of collection times and order the results in a meaningful way to identify peak usage periods.

In [34]:
%%sql

SELECT HOUR(time_of_record) AS hour_of_day,
    ROUND(AVG(time_in_queue)) AS avg_queue_time
FROM visits
WHERE NULLIF(time_of_record, 0) IS NOT NULL
GROUP BY hour_of_day
ORDER BY hour_of_day
LIMIT 5;


 * mysql+pymysql://root:***@localhost:3309/md_water_services
5 rows affected.


hour_of_day,avg_queue_time
6,149
7,149
8,149
9,49
10,48


In [35]:
%%sql

SELECT TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
    ROUND(AVG(time_in_queue)) AS avg_queue_time
FROM visits
WHERE NULLIF(time_of_record, 0) IS NOT NULL
GROUP BY hour_of_day
ORDER BY hour_of_day
LIMIT 5;


 * mysql+pymysql://root:***@localhost:3309/md_water_services
5 rows affected.


hour_of_day,avg_queue_time
06:00,149
07:00,149
08:00,149
09:00,49
10:00,48


In [36]:
%%sql

SELECT TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
    DAYNAME(time_of_record),
    CASE
        WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue
        ELSE NULL
    END AS Sunday
FROM visits
WHERE
    time_in_queue != 0
    LIMIT 5; -- this exludes other sources with 0 queue times.

 * mysql+pymysql://root:***@localhost:3309/md_water_services
5 rows affected.


hour_of_day,DAYNAME(time_of_record),Sunday
09:00,Friday,None
09:00,Friday,None
10:00,Friday,None
10:00,Friday,None
11:00,Friday,None


In [37]:
%%sql
-- This query calculates the average queue time for each hour of the day, broken down by day of the week (Sunday to Friday), and presents the results in a pivot-style format.

SELECT
TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
-- Sunday
ROUND(AVG(
CASE
    WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue
ELSE NULL
END
),0) AS Sunday,
-- Monday
ROUND(AVG(
CASE
WHEN DAYNAME(time_of_record) = 'Monday' THEN time_in_queue
ELSE NULL
END
),0) AS Monday,
-- Tuesday
ROUND(AVG(
CASE
WHEN DAYNAME(time_of_record) = 'Tuesday' THEN time_in_queue
ELSE NULL
END
),0) AS Tuesday,
-- Wednesday
ROUND(AVG(
CASE
WHEN DAYNAME(time_of_record) = 'Wednesday' THEN time_in_queue
ELSE NULL
END
),0) AS Wednesday,

-- Thurday
ROUND(AVG(
CASE
WHEN DAYNAME(time_of_record) = 'Thursday' THEN time_in_queue
ELSE NULL
END
),0) AS Thursday,

-- Friday
ROUND(AVG(
CASE
WHEN DAYNAME(time_of_record) = 'Friday' THEN time_in_queue
ELSE NULL
END
),0) AS Friday
FROM
visits
WHERE
time_in_queue != 0 -- this excludes other sources with 0 queue times
GROUP BY
hour_of_day
ORDER BY
hour_of_day;


 * mysql+pymysql://root:***@localhost:3309/md_water_services
14 rows affected.


hour_of_day,Sunday,Monday,Tuesday,Wednesday,Thursday,Friday
06:00,79,190,134,112,134,153
07:00,82,186,128,111,139,156
08:00,86,183,130,119,129,153
09:00,84,127,105,94,99,107
10:00,83,119,99,89,95,112
11:00,78,115,102,86,99,104
12:00,78,115,97,88,96,109
13:00,81,122,97,98,101,115
14:00,83,127,104,92,96,110
15:00,83,126,104,88,92,110


Here are the spotted patterns:
1. Queues are very long on a Monday morning and Monday evening as people rush to get water.
2. Wednesday has the lowest queue times, but long queues on Wednesday evening.
3. People have to queue pretty much twice as long on Saturdays compared to the weekdays. It looks like people spend their Saturdays queueing
for water, perhaps for the week's supply?
4. The shortest queues are on Sundays, and this is a cultural thing. The people of Maji Ndogo prioritise family and religion, so Sundays are spent
with family and friends.

16:36
We built a pivot table in SQL! The thing I want you to remember today is: SQL is a set of tools we can apply. By understanding CASE, we could
build a complex query that aggregates our data in a format that is very easy to understand.

#### Key Observed Patterns
The analysis reveals several consistent patterns in queue behaviour across days of the week:
1. **Monday peaks:** Queue times are significantly higher on Monday mornings and evenings, likely reflecting increased demand at the start of the week.
2. **Midweek variation:** Wednesdays generally exhibit lower queue times overall, although there is a noticeable increase during the evening period.
3. **Weekend pressure:** Queue times on Saturdays are substantially higher, approximately double those observed on weekdays, suggesting that households may be collecting water for extended use.
4. **Sunday decline:** Sundays show the lowest queue times, indicating reduced demand, potentially influenced by social or cultural factors affecting daily routines.
---
This analysis demonstrates how SQL can be used not only for data retrieval but also for transforming and structuring data into meaningful formats. By leveraging conditional aggregation through `CASE` statements, a pivot-style table was created to enable clear comparison of queue patterns across both time of day and day of the week.

## Water Crises in Maji_Ndogo Part2 Summary

This analysis evaluates water access across Maji Ndogo by examining the types of water sources used, the population served by each source, and the time required to access water. The objective is to identify key challenges in water accessibility and provide data-driven insights to inform decision-making.

### Key Insights

1. The majority of water sources are located in rural areas, indicating that access challenges are more pronounced outside urban centres.  
2. Approximately 43% of the population relies on shared taps, with an average of around 2,000 people served per tap, highlighting significant pressure on shared infrastructure.  
3. Around 31% of the population has water infrastructure installed within their homes; however, 45% of these systems are non-functional due to issues with supporting infrastructure such as pipes, pumps, and reservoirs.  
4. Approximately 18% of the population depends on wells, of which only 28% are classified as clean, indicating substantial water quality concerns.  
5. Access to water is time-intensive, with average queue times exceeding 120 minutes.  
6. Queue patterns reveal that:
   - Queue times are highest on Saturdays  
   - Demand peaks during mornings and evenings  
   - Wednesdays and Sundays generally experience lower queue times  

---

## Strategic Prioritisation Framework

Based on the analysis, a prioritised approach to intervention is proposed:

1. **Focus on high-impact water sources**
   - Improving shared taps should be prioritised, as they serve the largest proportion of the population  
   - Wells represent an important source but require quality improvements due to contamination  
   - Repairing existing infrastructure can significantly improve access and simultaneously reduce queue times  
   - Expanding in-home water access is resource-intensive and should be considered a longer-term objective  

2. **Address geographic challenges**
   - Since most water sources are located in rural areas, implementation strategies must account for logistical constraints such as limited infrastructure, supply access, and workforce availability  

---

## Recommended Interventions

The following practical interventions are proposed to address both short-term and long-term challenges:

1. **River-dependent communities**
   - Deploy water tankers as a temporary solution  
   - Develop long-term solutions through well construction  

2. **Well improvements**
   - Install appropriate filtration systems:
     - UV filtration for biological contamination  
     - Reverse osmosis for chemically polluted sources  
   - Investigate and address root causes of contamination  

3. **Shared tap optimisation**
   - Deploy additional water tankers to high-demand locations during peak periods  
   - Use queue time analysis to optimise timing and distribution  
   - Expand infrastructure by installing additional taps where needed  

   The long-term objective is to reduce queue times below the recommended threshold of 30 minutes.

4. **Low-queue shared taps**
   - Further improvements may require installation of in-home taps, which should be considered a long-term investment due to resource constraints  

5. **Infrastructure repair**
   - Prioritise repair of critical infrastructure (e.g., reservoirs, pipelines, pumping systems)  
   - Target interventions in areas where multiple water sources are affected to maximise impact  

---

## Conclusion

This analysis highlights significant challenges in water accessibility, infrastructure reliability, and service efficiency. By prioritising high-impact interventions and leveraging data-driven insights, it is possible to improve water access for a large portion of the population while optimising resource allocation.